README.md
# Heapster

***DEPRECATED***: Heapster is deprecated.  Consider using [metrics-server](https://github.com/kubernetes-incubator/metrics-server)
and a third party metrics pipeline to gather Prometheus-format metrics instead.  See the [deprecation timeline](docs/deprecation.md)
for more information on support. We will not be adding any new features to Heapster.

[![GoDoc](https://godoc.org/k8s.io/heapster?status.svg)](https://godoc.org/k8s.io/heapster) [![Build Status](https://travis-ci.org/kubernetes/heapster.svg?branch=master)](https://travis-ci.org/kubernetes/heapster)  [![Go Report Card](https://goreportcard.com/badge/github.com/kubernetes/heapster)](https://goreportcard.com/report/github.com/kubernetes/heapster)

Heapster enables Container Cluster Monitoring and Performance Analysis for [Kubernetes](https://github.com/kubernetes/kubernetes) (versions v1.0.6 and higher), and platforms which include it.

Heapster collects and interprets various signals like compute resource usage, lifecycle events, etc.
Note that the model API, formerly used provide REST access to its collected metrics, is now deprecated.
Please see [the model documentation](docs/model.md) for more details.

Heapster supports multiple sources of data.
More information [here](docs/source-configuration.md).

Heapster supports the pluggable storage backends described [here](docs/sink-owners.md).
We welcome patches that add additional storage backends.
Documentation on storage sinks [here](docs/sink-configuration.md).
The current version of Storage Schema is documented [here](docs/storage-schema.md).

### Running Heapster on Kubernetes

Heapster can run on a Kubernetes cluster using a number of backends.  Some common choices:
- [Metricly](docs/metricly.md)
- [Stackdriver Monitoring and Logging](docs/google.md) for Google Cloud Platform
- [Other backends](docs/)

### Running Heapster on OpenShift

Using Heapster to monitor an OpenShift cluster requires some additional changes to the Kubernetes instructions to allow communication between the Heapster instance and OpenShift's secured endpoints. To run standalone Heapster or a combination of Heapster and Hawkular-Metrics in OpenShift, follow [this guide](https://github.com/openshift/origin-metrics).

#### Troubleshooting guide [here](docs/debugging.md)


## Community

Contributions, questions, and comments are all welcomed and encouraged! Developers hang out on [Slack](https://kubernetes.slack.com) in the #sig-instrumentation channel (get an invitation [here](http://slack.kubernetes.io/)). We also have the [kubernetes-dev Google Groups mailing list](https://groups.google.com/forum/#!forum/kubernetes-dev). If you are posting to the list please prefix your subject with "heapster: ".


RELEASES.md
# Metrilcy Heapster Docker Container Release History
======================================

## Version 0.0.9
---------------------------
- default configuration no longer includes sys_containers

## Version 0.0.8
---------------------------
- fix a bug on negative computed cpu.usage.percent values
- fix a bug on 0 values for cpu.usage_rate for pod, namespace and cluster
- add a configurable cluster name in Metricly Sink to avoid element name collisions

## Version 0.0.7
---------------------------
- support creating computed metrics for cpu/memeory usage percent for all element types

## Version 0.0.6
---------------------------
- update default api uri to use kubernetes datasource
- fix element relationships broken by improved element type names

## Version 0.0.5
---------------------------
- improve element type names

## Version 0.0.4
---------------------------
- fix a bug when shortening a name that has no type prepended, i.e. 'cluster'

## Version 0.0.3
---------------------------
- add filter support to include/exclude metricsets by entity labels
- break concatenated custom labels to indivdual tags for better grouping elments
- shorten display element name by removing the types in its fqn

## Version 0.0.2
---------------------------
- add batch support to send element payload to Metricly

## Version 0.0.1
---------------------------
- initial release


common/metricly/metricly_test.go
// Copyright 2018 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package metricly

import (
	"net/url"
	"testing"
)

func TestParsingMetriclyConfig(t *testing.T) {
	//given
	metriclyURI := "https://api.app.metricly.com/ingest/kubernetes?cluster=qa&apiKey=datasourceApiKey&elementBatchSize=60&filter=label:{type:pod.*}&filter=!label:{type:pod_container}&metricCacheTTLSeconds=120"
	uri, err := url.Parse(metriclyURI)
	if err != nil {
		t.Fatalf("Error when parsing Metricly URL: %s", err.Error())
	}
	//when
	config, err := Config(uri)
	if err != nil {
		t.Fatalf("Error when configuring Metricly URL: %s", err.Error())
	}
	//then
	if config.ApiURL != "https://api.app.metricly.com/ingest/kubernetes" {
		t.Fatalf("The Api URL is wrong, actual=%s, expected=%s", config.ApiURL, "https://api.app.metricly.com/ingest/kubernetes")
	}
	if config.ApiKey != "datasourceApiKey" {
		t.Fatalf("The Api Key is wrong, actual=%s, expected=%s", config.ApiKey, "datasourceApiKey")
	}
	if config.ClusterName != "qa" {
		t.Fatalf("The cluster name is wrong, actual=%s, expected=%s", config.ClusterName, "qa")
	}
	if config.ElementBatchSize != 60 {
		t.Fatalf("The Element Batch Size is wrong, actual=%d, expected=%d", config.ElementBatchSize, 60)
	}
	if len(config.InclusionFilters) != 1 {
		t.Fatalf("There should be exactly 1 inclusion filter, actual=%d", len(config.InclusionFilters))
	}
	inFilter := config.InclusionFilters[0]
	if inFilter.Type != "label" || inFilter.Name != "type" || inFilter.Regex.String() != "pod.*" {
		t.Fatalf("The inclusion filter has the wrong value")
	}
	if len(config.ExclusionFilters) != 1 {
		t.Fatalf("There should be exactly 1 exclusion filter, actual=%d", len(config.ExclusionFilters))
	}
	exFilter := config.ExclusionFilters[0]
	if exFilter.Type != "label" || exFilter.Name != "type" || exFilter.Regex.String() != "pod_container" {
		t.Fatalf("The exclusion filter has the wrong value")
	}
	if config.MetricCacheTTLSeconds != 120 {
		t.Fatalf("The Metric Cache TTL is wrong, actual=%d, expected=%d", config.MetricCacheTTLSeconds, 120)
	}
}


metrics/sinks/metricly/metricly_test.go
// Copyright 2018 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package metricly

import (
	"regexp"
	"testing"
	"time"

	metricly_core "github.com/metricly/go-client/model/core"
	"k8s.io/heapster/common/metricly"
	"k8s.io/heapster/metrics/core"
)

func TestShortenNameWithoutType(t *testing.T) {
	//given
	fqn := "cluster"
	//when & then
	if shortenName(fqn) != "cluster" {
		t.Errorf("Shorten fqn failed on single name with no type")
	}
}

func TestPrettyElementType(t *testing.T) {
	if prettyElementType("cluster") != "Kubernetes Cluster" {
		t.Errorf("pretty element type for 'cluster' failed to match expected 'Kubernetes Cluster'")
	}
	if prettyElementType("ns") != "Kubernetes Namespace" {
		t.Errorf("pretty element type for 'ns' failed to match expected 'Kubernetes Namespace'")
	}
	if prettyElementType("node") != "Kubernetes Node" {
		t.Errorf("pretty element type for 'node' failed to match expected 'Kubernetes Node'")
	}
	if prettyElementType("pod") != "Kubernetes Pod" {
		t.Errorf("pretty element type for 'pod' failed to match expected 'Kubernetes Pod'")
	}
	if prettyElementType("pod_container") != "Kubernetes Pod Container" {
		t.Errorf("pretty element type for 'pod_container' failed to match expected 'Kubernetes Pod Container'")
	}
	if prettyElementType("sys_container") != "Kubernetes Sys Container" {
		t.Errorf("pretty element type for 'sys_container' failed to match expected 'Kubernetes Sys Container'")
	}
}
func TestConvertDataBatchToElements(t *testing.T) {
	//given
	batch := createDataBatch()
	//when
	elements := DataBatchToElements("", metricly.MetriclyConfig{}, NewMetricCache(300), batch)
	//then
	if len(elements) != 1 {
		t.Errorf("There should be 1 element in elements, but actual =  %d", len(elements))
	}
	e := elements[0]
	if e.Id != "namespace:kube-system/pod:heapster-ww9r6" {
		t.Errorf("The element id is wrong")
	}
	if e.Name != "kube-system/heapster-ww9r6" {
		t.Errorf("The element name is not propertly shortened")
	}
	if len(e.Tags()) != 10 {
		t.Errorf("The element has the wrong number of tags")
	}
	tagMode, found := e.Tag("addonmanager.kubernetes.io/mode")
	if !found || tagMode.Value != "Reconcile" {
		t.Errorf("Custom Label 'addonmanager.kubernetes.io/mode' is not converted to an expected element tag")
	}
	tagApp, found := e.Tag("k8s-app")
	if !found || tagApp.Value != "heapster" {
		t.Errorf("Custom Label 'k8s-app' is not converted to an expected element tag")
	}
	if len(e.Metrics()) != 2 {
		t.Errorf("The element has the wrong number of metrics")
	}
	if len(e.Samples()) != 2 {
		t.Errorf("The element has the wrong number of samples")
	}

}

func createDataBatch() *core.DataBatch {
	timestamp, _ := time.Parse(time.RFC3339, "2018-01-01T00:00:00+00:00")
	batch := core.DataBatch{
		Timestamp:  timestamp,
		MetricSets: make(map[string]*core.MetricSet),
	}

	batch.MetricSets["namespace:kube-system/pod:heapster-ww9r6"] = &core.MetricSet{
		Labels: map[string]string{
			"host_id":        "minikube",
			"hostname":       "minikube",
			"labels":         "addonmanager.kubernetes.io/mode:Reconcile,k8s-app:heapster",
			"namespace_id":   "efc44650-ffbd-11e7-a0bf-080027b171ba",
			"namespace_name": "kube-system",
			"nodename":       "minikube",
			"pod_id":         "761fc05e-008c-11e8-911d-080027b171ba",
			"pod_name":       "heapster-ww9r6",
			"type":           "pod",
		},
		MetricValues: map[string]core.MetricValue{
			"cpu/usage_rate": {
				MetricType: core.MetricGauge,
				ValueType:  core.ValueInt64,
				IntValue:   9248344,
			},
		},
		LabeledMetrics: []core.LabeledMetric{
			{
				Name: "filesystem/usage",
				MetricValue: core.MetricValue{
					MetricType: core.MetricGauge,
					ValueType:  core.ValueInt64,
					IntValue:   835584,
				},
				Labels: map[string]string{
					"resource_id": "/dev/sda1",
				},
			},
		},
	}

	return &batch
}

func TestInclusionFilter(t *testing.T) {
	//given
	re, _ := regexp.Compile("pod.*")
	filters := []metricly.Filter{
		{
			Type:  "label",
			Name:  "type",
			Regex: re,
		},
	}
	ms := &core.MetricSet{
		Labels: map[string]string{
			"type": "pod",
		},
	}
	//when & then
	if !include(filters, ms) {
		t.Errorf("The metric set should be included and return true")
	}
}

func TestExclusionFilter(t *testing.T) {
	re, _ := regexp.Compile("sys_.*")
	filters := []metricly.Filter{
		{
			Type:  "label",
			Name:  "type",
			Regex: re,
		},
	}
	ms := &core.MetricSet{
		Labels: map[string]string{
			"type": "sys_container",
		},
	}
	if !exclude(filters, ms) {
		t.Errorf("The metric set should be excluded and return true")
	}
}

func TestFilter(t *testing.T) {
	//given
	ire, _ := regexp.Compile("pod.*")
	ifilters := []metricly.Filter{
		{
			Type:  "label",
			Name:  "type",
			Regex: ire,
		},
	}
	ere, _ := regexp.Compile("pod_container")
	efilters := []metricly.Filter{
		{
			Type:  "label",
			Name:  "type",
			Regex: ere,
		},
	}
	pod := &core.MetricSet{
		Labels: map[string]string{
			"type": "pod",
		},
	}
	container := &core.MetricSet{
		Labels: map[string]string{
			"type": "pod_container",
		},
	}
	//when & then
	if !filter(ifilters, efilters, pod) {
		t.Errorf("pod should be pass the filter")
	}
	if filter(ifilters, efilters, container) {
		t.Errorf("pod container should not pass the filter")
	}
}

func TestLinkElements(t *testing.T) {
	//given
	elements := createElements()
	//when
	LinkElements(elements)
	//then
	if len(elements) != 6 {
		t.Errorf("There should be 6 elements in elements, but actual =  %d", len(elements))
	}
	cluster := elements[0]
	podContainer := elements[1]
	pod := elements[2]
	node := elements[3]
	ns := elements[4]
	sysContainer := elements[5]
	if len(cluster.Relations()) != 0 {
		t.Errorf("The cluster element should have no relations")
	}
	if len(podContainer.Relations()) != 0 {
		t.Errorf("The pod_container element should have no relations")
	}
	if len(pod.Relations()) != 1 {
		t.Errorf("The pod element should have 1 relation, but only has %d", len(pod.Relations()))
	}
	if len(node.Relations()) != 3 {
		t.Errorf("The node elment should have 3 relations, but only has %d", len(node.Relations()))
	}
	if len(ns.Relations()) != 2 {
		t.Errorf("The ns element should have 2 relations, but only has %d", len(ns.Relations()))
	}
	if len(sysContainer.Relations()) != 0 {
		t.Errorf("The sys_container element should have no relations")
	}
}

func createElements() []metricly_core.Element {
	//cluster
	cluster := metricly_core.NewElement("cluster", "cluster", "Kubernetes Cluster", "")
	cluster.AddTag("type", "cluster")

	//pod_container
	podContainer := metricly_core.NewElement("namespace:kube-system/pod:heapster-ww9r6/container:heapster", "namespace:kube-system/pod:heapster-ww9r6/container:heapster", "Kubernetes Pod Container", "")
	podContainer.AddTag("type", "pod_container")
	podContainer.AddTag("pod_id", "761fc05e-008c-11e8-911d-080027b171ba")
	podContainer.AddTag("namespace_id", "efc44650-ffbd-11e7-a0bf-080027b171ba")
	podContainer.AddTag("host_id", "minikube")

	//pod
	pod := metricly_core.NewElement("namespace:kube-system/pod:heapster-ww9r6", "namespace:kube-system/pod:heapster-ww9r", "Kubernetes Pod", "")
	pod.AddTag("type", "pod")
	pod.AddTag("pod_id", "761fc05e-008c-11e8-911d-080027b171ba")
	pod.AddTag("host_id", "minikube")
	pod.AddTag("namespace_id", "efc44650-ffbd-11e7-a0bf-080027b171ba")

	//node
	node := metricly_core.NewElement("node:minikube", "node:minikube", "Kubernetes Node", "")
	node.AddTag("type", "node")
	node.AddTag("host_id", "minikube")

	//namespace
	ns := metricly_core.NewElement("namespace:kube-system", "namespace:kube-system", "Kubernetes Namespace", "")
	ns.AddTag("type", "ns")
	ns.AddTag("namespace_id", "efc44650-ffbd-11e7-a0bf-080027b171ba")

	//sys_container
	sysContainer := metricly_core.NewElement("node:minikube/container:init.scope", "node:minikube/container:init.scope", "Kubernetes Sys Container", "")
	sysContainer.AddTag("type", "sys_container")
	sysContainer.AddTag("host_id", "minikube")

	return []metricly_core.Element{cluster, podContainer, pod, node, ns, sysContainer}
}


metrics/sinks/factory.go
// Copyright 2015 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package sinks

import (
	"fmt"
	"time"

	"github.com/golang/glog"
	"k8s.io/heapster/common/flags"
	"k8s.io/heapster/metrics/core"
	"k8s.io/heapster/metrics/sinks/elasticsearch"
	"k8s.io/heapster/metrics/sinks/gcm"
	"k8s.io/heapster/metrics/sinks/graphite"
	"k8s.io/heapster/metrics/sinks/hawkular"
	"k8s.io/heapster/metrics/sinks/honeycomb"
	"k8s.io/heapster/metrics/sinks/influxdb"
	"k8s.io/heapster/metrics/sinks/kafka"
	"k8s.io/heapster/metrics/sinks/librato"
	logsink "k8s.io/heapster/metrics/sinks/log"
	metricsink "k8s.io/heapster/metrics/sinks/metric"
	"k8s.io/heapster/metrics/sinks/metricly"
	"k8s.io/heapster/metrics/sinks/opentsdb"
	"k8s.io/heapster/metrics/sinks/riemann"
	"k8s.io/heapster/metrics/sinks/stackdriver"
	"k8s.io/heapster/metrics/sinks/statsd"
	"k8s.io/heapster/metrics/sinks/wavefront"
)

type SinkFactory struct {
}

func (this *SinkFactory) Build(uri flags.Uri) (core.DataSink, error) {
	switch uri.Key {
	case "elasticsearch":
		return elasticsearch.NewElasticSearchSink(&uri.Val)
	case "gcm":
		return gcm.CreateGCMSink(&uri.Val)
	case "stackdriver":
		return stackdriver.CreateStackdriverSink(&uri.Val)
	case "statsd":
		return statsd.NewStatsdSink(&uri.Val)
	case "graphite":
		return graphite.NewGraphiteSink(&uri.Val)
	case "hawkular":
		return hawkular.NewHawkularSink(&uri.Val)
	case "influxdb":
		return influxdb.CreateInfluxdbSink(&uri.Val)
	case "kafka":
		return kafka.NewKafkaSink(&uri.Val)
	case "librato":
		return librato.CreateLibratoSink(&uri.Val)
	case "log":
		return logsink.NewLogSink(), nil
	case "metric":
		return metricsink.NewMetricSink(140*time.Second, 15*time.Minute, []string{
			core.MetricCpuUsageRate.MetricDescriptor.Name,
			core.MetricMemoryUsage.MetricDescriptor.Name}), nil
	case "opentsdb":
		return opentsdb.CreateOpenTSDBSink(&uri.Val)
	case "wavefront":
		return wavefront.NewWavefrontSink(&uri.Val)
	case "riemann":
		return riemann.CreateRiemannSink(&uri.Val)
	case "honeycomb":
		return honeycomb.NewHoneycombSink(&uri.Val)
	case "metricly":
		return metricly.NewMetriclySink(&uri.Val)
	default:
		return nil, fmt.Errorf("Sink not recognized: %s", uri.Key)
	}
}

func (this *SinkFactory) BuildAll(uris flags.Uris, historicalUri string, disableMetricSink bool) (*metricsink.MetricSink, []core.DataSink, core.HistoricalSource) {
	result := make([]core.DataSink, 0, len(uris))
	var metric *metricsink.MetricSink
	var historical core.HistoricalSource
	for _, uri := range uris {
		sink, err := this.Build(uri)
		if err != nil {
			glog.Errorf("Failed to create %v sink: %v", uri, err)
			continue
		}
		if uri.Key == "metric" {
			metric = sink.(*metricsink.MetricSink)
		}
		if uri.String() == historicalUri {
			if asHistSource, ok := sink.(core.AsHistoricalSource); ok {
				historical = asHistSource.Historical()
			} else {
				glog.Errorf("Sink type %q does not support being used for historical access", uri.Key)
			}
		}
		result = append(result, sink)
	}

	if len([]flags.Uri(uris)) != 0 && len(result) == 0 {
		glog.Fatal("No available sink to use")
	}

	if metric == nil && !disableMetricSink {
		uri := flags.Uri{}
		uri.Set("metric")
		sink, err := this.Build(uri)
		if err == nil {
			result = append(result, sink)
			metric = sink.(*metricsink.MetricSink)
		} else {
			glog.Errorf("Error while creating metric sink: %v", err)
		}
	}
	if len(historicalUri) > 0 && historical == nil {
		glog.Errorf("Error while initializing historical access: unable to use sink %q as a historical source", historicalUri)
	}
	return metric, result, historical
}

func NewSinkFactory() *SinkFactory {
	return &SinkFactory{}
}


hooks/check_boilerplate.sh
#!/bin/bash

# Copyright 2015 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

REF_FILE="./hooks/boilerplate.go.txt"
if [ ! -e $REF_FILE ]; then
  echo "Missing reference file: " ${REF_FILE}
  exit 1
fi

LINES=$(cat "${REF_FILE}" | wc -l | tr -d ' ')
GO_FILES=$(find . -name "*.go" | grep -v -e "vendor")

for FILE in ${GO_FILES}; do
  DIFFER=$(cat "${FILE}" | sed 's/201[5-9]\{1\}/2014/g' | head "-${LINES}" | diff -q - "${REF_FILE}")

  if [[ ! -z "${DIFFER}" ]]; then
    echo "${FILE} does not have the correct copyright notice."
    exit 1
  fi
done


metrics/sinks/metricly/metric_cache.go
// Copyright 2018 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package metricly

import (
	"sync"
	"time"

	"github.com/metricly/go-client/model/core"
)

type metricKey struct {
	elementId string
	metricId  string
}

type sample struct {
	timestamp int64
	val       float64
}

//MetricCache contains a collection of samples for COUNTER metrics
// counters contains predefined metrics that are allowed to be added to the cache
type MetricCache struct {
	counters map[string]struct{}
	cache    map[metricKey]sample
	lock     sync.Mutex
}

//NewMetricCache creates a new MetricCache
func NewMetricCache(ttl int) (mc *MetricCache) {
	mc = &MetricCache{cache: make(map[metricKey]sample),
		counters: map[string]struct{}{
			"cpu.usage": {},
		}}
	go func() {
		duration := time.Duration(ttl) * time.Second
		for now := range time.Tick(duration) {
			mc.lock.Lock()
			for k, v := range mc.cache {
				if now.Unix()*1000-v.timestamp > int64(ttl*1000) {
					delete(mc.cache, k)
				}
			}
			mc.lock.Unlock()
		}
	}()
	return
}

// returns the size of the metric cache
func (mc *MetricCache) Size() int {
	mc.lock.Lock()
	size := len(mc.cache)
	mc.lock.Unlock()
	return size
}

func (mc *MetricCache) ContainsMetric(metricId string) bool {
	_, found := mc.counters[metricId]
	return found
}

func (mc *MetricCache) addSample(elementId string, s core.Sample) {
	if _, ok := mc.counters[s.MetricId()]; ok {
		mc.lock.Lock()
		mc.cache[metricKey{elementId, s.MetricId()}] = sample{s.Timestamp(), s.Val()}
		mc.lock.Unlock()
	}
}

func (mc *MetricCache) getSample(elementId, metricId string) (sample, bool) {
	mc.lock.Lock()
	s, ok := mc.cache[metricKey{elementId, metricId}]
	mc.lock.Unlock()
	return s, ok
}


metrics/sinks/metricly/metricly.go
// Copyright 2018 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package metricly

import (
	"net/url"
	"strings"

	"github.com/golang/glog"
	"github.com/metricly/go-client/api"
	metricly_core "github.com/metricly/go-client/model/core"
	"k8s.io/heapster/common/metricly"
	"k8s.io/heapster/metrics/core"
)

const (
	infraContainerName           = "POD"
	defaultElementsPayloadSize   = 20
	defaultMetricCacheTTLSeconds = 300
	elementTypePrefix            = "Kubernetes "
	defaultClusterName           = "k8s-cluster"
)

type MetriclyMetricsSink struct {
	client  api.Client
	config  metricly.MetriclyConfig
	cache   *MetricCache
	cluster string
}

func (sink *MetriclyMetricsSink) Name() string {
	return "Metricly Metrics Sink"
}

func (sink *MetriclyMetricsSink) Stop() {
}

type chunk struct {
	start, end int
}

func (sink *MetriclyMetricsSink) ExportData(batch *core.DataBatch) {
	glog.Info("Start exporting data batch to Metricly ...")
	elements := DataBatchToElements(sink.cluster+"/", sink.config, sink.cache, batch)
	elementsPayloadSize := defaultElementsPayloadSize
	if sink.config.ElementBatchSize > 0 {
		elementsPayloadSize = sink.config.ElementBatchSize
	}
	total := len(elements)
	chunks := partition(total, elementsPayloadSize)
	jobs := make(chan chunk, len(chunks))
	count := make(chan int)
	for _, c := range chunks {
		jobs <- c
	}
	close(jobs)

	for c := range jobs {
		go func(c chunk) {
			if err := sink.send(elements[c.start:c.end]); err == nil {
				count <- c.end - c.start
			} else {
				glog.Warningf("Error occurred during exporting %d elements with response:  %v", c.end-c.start, err)
				count <- 0
			}
		}(c)
	}

	var sent int
	for i := 0; i < len(chunks); i++ {
		sent += <-count
	}
	glog.Infof("Exported %d out of %d elements using %d workers", sent, len(elements), len(chunks))
}

func NewMetriclySink(uri *url.URL) (core.DataSink, error) {
	config, _ := metricly.Config(uri)
	glog.Info("Create Metricly sink using config: ", config)
	mcttl := defaultMetricCacheTTLSeconds
	if config.MetricCacheTTLSeconds > 0 {
		mcttl = config.MetricCacheTTLSeconds
	}
	clusterName := defaultClusterName
	if len(config.ClusterName) > 0 {
		clusterName = config.ClusterName
	}
	return &MetriclyMetricsSink{client: api.NewClient(config.ApiURL, config.ApiKey), config: config, cache: NewMetricCache(mcttl), cluster: clusterName}, nil
}

func DataBatchToElements(cluster string, config metricly.MetriclyConfig, cache *MetricCache, batch *core.DataBatch) []metricly_core.Element {
	ts := batch.Timestamp.Unix() * 1000
	var elements []metricly_core.Element
	for key, ms := range batch.MetricSets {
		if !filter(config.InclusionFilters, config.ExclusionFilters, ms) {
			glog.V(1).Info("metric set is dropped due to filtering, key: ", key)
			continue
		}
		if strings.HasSuffix(key, infraContainerName) {
			glog.V(1).Info("metric set is dropped due to it is a infra/pause container: " + key)
			continue
		}
		etype := ms.Labels["type"]
		element := metricly_core.NewElement(cluster+key, cluster+shortenName(key), prettyElementType(etype), "")
		// metric set labels to element tags
		for lname, lvalue := range ms.Labels {
			if lname == "labels" {
				for _, l := range strings.Split(lvalue, ",") {
					if strings.TrimSpace(l) == "" {
						continue
					}
					kv := strings.SplitN(l, ":", 2)
					element.AddTag(kv[0], kv[1])
				}
			} else {
				element.AddTag(lname, lvalue)
			}
		}
		// metrics
		for mname, mvalue := range ms.MetricValues {
			if sample, err := metricly_core.NewSample(sanitizeMetricId(mname), ts, mvalue.GetValue()); err == nil {
				if cache.ContainsMetric(sample.MetricId()) {
					if lastSample, found := cache.getSample(element.Id, sample.MetricId()); found {
						if delta, err := metricly_core.NewSample(sample.MetricId(), ts, sample.Val()-lastSample.val); err == nil {
							element.AddSample(delta)
						}
					}
					cache.addSample(element.Id, sample)
				} else {
					element.AddSample(sample)
				}
			}
		}
		// labeled metrics
		for _, lmetric := range ms.LabeledMetrics {
			instanceMetricName := sanitizeMetricId(lmetric.Name) + ":" + lmetric.Labels["resource_id"]
			if sample, err := metricly_core.NewSample(instanceMetricName, ts, lmetric.GetValue()); err == nil {
				element.AddSample(sample)
			}
		}
		elements = append(elements, element)
	}
	LinkElements(elements)
	CreateComputedMetrics(ts, elements)
	return elements
}

func LinkElements(elements []metricly_core.Element) {
	var elementsById = make(map[string]*metricly_core.Element)
	for idx := range elements {
		switch e := elements[idx]; e.Type {
		case "Kubernetes Pod":
			if id, ok := elements[idx].Tag("pod_id"); ok {
				elementsById[id.Value] = &e
			}
		case "Kubernetes Node":
			if id, ok := elements[idx].Tag("host_id"); ok {
				elementsById[id.Value] = &e
			}
		case "Kubernetes Namespace":
			if id, ok := elements[idx].Tag("namespace_id"); ok {
				elementsById[id.Value] = &e
			}
		}
	}

	for idx := range elements {
		switch e := elements[idx]; e.Type {
		case "Kubernetes Pod Container":
			if podId, ok := e.Tag("pod_id"); ok {
				if pod, ok := elementsById[podId.Value]; ok {
					pod.AddRelation(e.Id)
				}
			}
			if hostId, ok := elements[idx].Tag("host_id"); ok {
				if host, ok := elementsById[hostId.Value]; ok {
					host.AddRelation(e.Id)
				}
			}
			if nsId, ok := elements[idx].Tag("namespace_id"); ok {
				if ns, ok := elementsById[nsId.Value]; ok {
					ns.AddRelation(e.Id)
				}
			}
		case "Kubernetes Pod":
			if hostId, ok := elements[idx].Tag("host_id"); ok {
				if host, ok := elementsById[hostId.Value]; ok {
					host.AddRelation(e.Id)
				}
			}
			if nsId, ok := elements[idx].Tag("namespace_id"); ok {
				if ns, ok := elementsById[nsId.Value]; ok {
					ns.AddRelation(e.Id)
				}
			}
		case "Kubernetes Sys Container":
			if hostId, ok := elements[idx].Tag("host_id"); ok {
				if host, ok := elementsById[hostId.Value]; ok {
					host.AddRelation(e.Id)
				}
			}
		}
	}
}

//CreateComputedMetrics creates new computed metrics/samples based on current collected metrics
func CreateComputedMetrics(timestamp int64, elements []metricly_core.Element) {
	for _, e := range elements {
		CreateComputedMetric(timestamp, &e)
	}
}

func shortenName(fqn string) string {
	var names []string
	for _, s := range strings.Split(fqn, "/") {
		kv := strings.SplitN(s, ":", 2)
		if len(kv) == 2 {
			names = append(names, kv[1])
		} else {
			names = append(names, kv[0])
		}
	}
	return strings.Join(names, "/")
}

func prettyElementType(etype string) string {
	switch etype {
	case "ns":
		return elementTypePrefix + "Namespace"
	}
	return elementTypePrefix + strings.Title(strings.Replace(etype, "_", " ", -1))
}

//filter MetricSet against inclusion/exclusion filters and return true if it passes
func filter(inf []metricly.Filter, exf []metricly.Filter, ms *core.MetricSet) bool {
	return include(inf, ms) && !exclude(exf, ms)
}

func exclude(filters []metricly.Filter, ms *core.MetricSet) bool {
	if len(filters) == 0 {
		return false
	}
	for k, v := range ms.Labels {
		for _, f := range filters {
			if f.Type == "label" && k == f.Name && f.Regex.MatchString(v) {
				return true
			}
		}
	}
	return false
}

func include(filters []metricly.Filter, ms *core.MetricSet) bool {
	if len(filters) == 0 {
		return true
	}
	for k, v := range ms.Labels {
		for _, f := range filters {
			if f.Type == "label" && k == f.Name && f.Regex.MatchString(v) {
				return true
			}
		}
	}
	return false
}

func sanitizeMetricId(metricId string) string {
	return strings.Replace(metricId, "/", ".", -1)
}

func partition(total, batch int) []chunk {
	partitions := total / batch
	var chunks []chunk
	var i int
	for i = 0; i < partitions; i++ {
		chunks = append(chunks, chunk{start: i * batch, end: (i + 1) * batch})
	}
	if total%batch != 0 {
		chunks = append(chunks, chunk{start: i * batch, end: total})
	}
	return chunks
}

func (sink *MetriclyMetricsSink) send(elements []metricly_core.Element) error {
	return sink.client.PostElements(elements)
}
